[dev-launcher] add button to clear recently opened apps, update UI to…

ajsmth opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

The recently opened apps list isn't terribly useful as it doesn't show enough information. This PR adds some extra context to these apps: 

- a local packager will include the packager url 
- an eas update will show the branch and update if loaded from the Extensions UI 
- an eas update will show the branch name if loaded manually from the url text input

# How

<!--
How did you build this feature or fix this bug and why?
-->

- updated the interface to pass the manifest to the recently opened apps registry
- return an array instead of a dictionary when loading cached apps from the registry (this should keep the ordering consistent in the list)
- parse out relevant app data from the manifest (e.g name and metadata)
- pass the update message to the registry via `?updateMessage=${update.message}` as a query when opening an update from the extensions UI



# Test Plan

This image shows the 4 variations of apps that might be visible:

The first is an update loaded via the extension UI
The second is a local packager loaded w/ `expo-cli` 
The third is an update loaded via the text input
The fourth is a bare app w/ the  `react-native` packager

<img width="569" alt="Screen Shot 2022-05-30 at 10 21 26 AM" src="https://user-images.githubusercontent.com/40680668/171038126-e24f8966-c6f9-4000-ba29-0ebf58ea1428.png">


I've tried to come up with all of the possible scenarios in regards to handling deep links and urls and its actually surprising how many different combinations we are handling here. Hopefully this has some guidance for testing this PR and future QA as I often struggle to wrap my head around all the scenarios and expected behaviours

URLs: 

- dev client deep link (`scheme://expo-development-client?url=...`)
- non dev client deep link (`scheme://home?...`)
- eas update url (`https://u.expo.dev/update/...`)
- dev client deep link + eas update url (`scheme://expo-development-client?url=https://u.expo.dev/update...`)
- dev client deep link + eas update url and message (`scheme://expo-development-client?url=https://u.expo.dev/update...&updateMessage=`)
- packager url (`http://localhost:8081`)

Modes of delivery: 
- in app url input 
- in app EAS Updates UI
- deep link while app is backgrounded
- deep link while app is closed
 

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

ajsmth committed:
[dev-launcher] add button to clear recently opened apps, update UI to include more information about the app'

- add button to clear recently opened apps
- update native implementation to work with objects to pass more data through to "recently opened app" buttons
- update UI to include more information about the app session

ajsmth committed:
add eas update to recent apps

ajsmth committed:
refactor

ajsmth committed:
revert uncomment

ajsmth committed:
[skip ci] changelog

esamelson reviewed (commented):
Thanks for taking care of this, the screenshot looks so much nicer!

I think that passing information around by mutating the update URL is brittle and might make things more confusing in the future. If we do go this route, I think it does make sense to strip the extra parameters before actually making a request to the URL -- better to send exactly what we need and not more.

One possible alternate approach could be to use deep-link-style URLs to open EAS updates from the UI. We could easily include extra query params in that kind of URL without mutating the actual manifest URL, i.e. `scheme://expo-development-client/?url=<url-encoded-manifest-url>&updateMessage=blah`. And that still leaves the door open for using QR code links to provide this information if we want to go that route, since they already use this deep link format.

ajsmth commented on packages/expo-dev-launcher/ios/EXDevLauncherRecentlyOpenedAppsRegistry.swift:
```diff
@@ -18,33 +19,70 @@ public class EXDevLauncherRecentlyOpenedAppsRegistry: NSObject {
+    if let manifest = manifest {
+      appEntry["name"] = manifest.name()
+      
+      let json = manifest.rawManifestJSON()
```
not for the metadata portion unless we update the manifest package

esamelson replied:
Gotcha. I think that would be preferred (since the main point of the manifests package is to insulate everything else from the exact object structure) but could easily be a separate follow-up task.

ajsmth commented:
> One possible alternate approach could be to use deep-link-style URLs to open EAS updates from the UI. We could easily include extra query params in that kind of URL without mutating the actual manifest URL, i.e. scheme://expo-development-client/?url=<url-encoded-manifest-url>&updateMessage=blah. And that still leaves the door open for using QR code links to provide this information if we want to go that route, since they already use this deep link format.

I think this is a great idea - I'll update the implementation

ajsmth committed:
update implementation around handling eas update urls to include additional params

ajsmth committed:
revert devsettings on bridge

ajsmth committed:
fix iOS tests broken by updated implementation

ajsmth committed:
cleanup impl

ajsmth committed:
fix ordering, reusing matchign entrys in cache on reload

ajsmth commented on packages/expo-dev-launcher/bundle/screens/HomeScreen.tsx:
```diff
@@ -254,48 +258,164 @@ function DevSessionList({ devSessions = [], onDevSessionPress }: DevSessionListP
+
+      <Spacer.Vertical size="small" bg="default" />
+
+      {Boolean(message) && (
```
this is now done - I think it looks good 

I added some logic to reuse the EAS Update information if it already exists in the cache (e.g when a user presses an update link in this list) otherwise we'll show the app name and update url

ajsmth commented on packages/expo-dev-launcher/bundle/providers/RecentlyOpenedAppsProvider.tsx:
```diff
@@ -37,14 +50,17 @@ export function useRecentlyOpenedApps() {
+        const recentApps = {};
+
+        for (const app of apps) {
+          // index by url to eliminate multiple bundlers with the same address
```
this is updated now to have the most recently opened apps first in the list

esamelson replied:
sorry to harp on this -- i'm thinking specifically about the situation where app A and app B have the same URL, and app B was opened more recently, but app A is reached later in this `for` loop and so it overrides app B in the `recentApps` object. but maybe i'm missing something and that's already handled?

ajsmth committed:
Merge branch 'main' into andrew/eng-4900-recently-opened-items

ajsmth commented:
I've updated the implementation with some of the suggestions made -- I've also changed some of the logic on the native side around loading apps

There are four ways to open an app via url:
- deep link
- manual input
- pressing on a packager
- pressing on a eas update 

all of these would at some point use some of the url validation helper methods in various different places to format the url to be loaded via `loadApp`, so I've moved this logic into `loadApp` itself, and created a helper class that formats and groups the url and queryParams for use in the `RecentlyOpenedAppsRegistry`

ajsmth commented on packages/expo-dev-launcher/bundle/hooks/useOnUpdatePress.tsx:
```diff
@@ -1,19 +1,20 @@
 export function useOnUpdatePress() {
   const toastStack = useToastStack();
-  const { runtimeVersion, updatesUrl } = useUpdatesConfig();
+  const { runtimeVersion, projectUrl } = useUpdatesConfig();
```
this was a bug that went unnoticed since the iOS module uses `Expo.plist` when this value is missing

ajsmth committed:
fix tests

esamelson commented on packages/expo-dev-launcher/bundle/hooks/useOnUpdatePress.tsx:
```diff
@@ -1,19 +1,20 @@
 
   const onUpdatePress = React.useCallback(
-    (update: Update) => {
+    async (update: Update) => {
```
```suggestion
    (update: Update) => {
```
no longer needed?

esamelson reviewed (approved):
Looks great, just one more tiny nit that may not be an issue (?)

ajsmth committed:
Update packages/expo-dev-launcher/bundle/hooks/useOnUpdatePress.tsx

Co-authored-by: Eric Samelson <esamelson@users.noreply.github.com>

ajsmth committed:
update logic for sorting by timestamp

ajsmth committed:
nit

ajsmth commented:
going to hold off on merging this as the Android portion won't be ready for release and we're planning on pushing out an update tomorrow morning

ajsmth committed:
[android][dev-launcher] update recently opened items on android

ajsmth commented:
@esamelson  @lukmccall  I've added the android implementation on the latest commit - I can split it out of this PR if you find that easier - let me know

lukmccall commented on packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherRecentlyOpenedAppsRegistry.kt:
```diff
@@ -4,38 +4,81 @@ import android.content.Context
+    appEntry["timestamp"] = timestamp
+    appEntry["url"] = url
+
+
```
```suggestion
```

lukmccall commented on packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/launcher/DevLauncherRecentlyOpenedAppsRegistry.kt:
```diff
@@ -4,38 +4,81 @@ import android.content.Context
+      val previousEntry = Gson().fromJson(previousEntryJsonString, Map::class.java)
+      appEntry = previousEntry.toMutableMap() as MutableMap<String, Any>
+    }
+
```
```suggestion
```

lukmccall reviewed (approved):
Looks fine, you don't have to split it ;)

ajsmth committed:
Update packages/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/DevLauncherController.kt

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

ajsmth committed:
Apply suggestions from code review

Co-authored-by: Łukasz Kosmaty <lukasz.kosmaty@swmansion.com>

esamelson commented on packages/expo-dev-launcher/android/src/debug/java/expo/modules/devlauncher/modules/DevLauncherInternalModule.kt:
```diff
@@ -118,13 +112,8 @@ class DevLauncherInternalModule(reactContext: ReactApplicationContext?) :
       try {
-        val parsedUrl = Uri.parse(url.trim())
-        val appUrl = if (isDevLauncherUrl(parsedUrl)) {
-          requireNotNull(getAppUrlFromDevLauncherUrl(parsedUrl)) { "The provided url doesn't contain the app url." }
```
Since this particular check has been received, could you make sure that this case (deep-link-style URL with no `url` query param) is part of the test plan and still shows a useful error message?

ajsmth replied:
@esamelson @lukmccall  I've updated the initial PR description to include some (hopefully all) of the different scenarios for handling links - can you have a look and see if there is anything missing? I will add this to the release guide as well based on feedback, and of course test these on this PR as well

esamelson replied:
Awesome, looks comprehensive! The only thing I would suggest is adding a few malformed URLs as well, to make sure our error messages/states are correct -- maybe things like `scheme://expo-development-client` or `scheme://expo-development-client/?url=blah`.

esamelson reviewed (approved):
I think this looks good, the main thing is just making sure that all the relevant cases are part of the test plan, since this PR has grown in scope.

ajsmth committed:
Merge branch 'main' into andrew/eng-4900-recently-opened-items

ajsmth committed:
Merge branch 'main' into andrew/eng-4900-recently-opened-items

ajsmth committed:
fix android build

ajsmth committed:
remove legacy test
